### DISCUSSION TITLE
How to sign-in as a certain user, under emulation?

### DISCUSSION THREAD

**Discussion by akauppi** [firebase/firebase-js-sdk#4547] [Q&A] [answered]
Under auth emulation, `signInWithCustomToken` used to allow forcing a sign-in as a certain user, simply by providing the token `{ "uid": "xyz" }`. This does not seem to work with 0.900.15 client.

Is this intentional?  What should I use, instead?

Attached code snippets, also [online](https://github.com/akauppi/GroundLevel-es-firebase/blob/recamp-firebase-9x/packages/app/vitebox/init/main.js#L110).

```
const app= initializeApp( {
    projectId,
    apiKey: "none",
    authDomain: "no.such.com"   // tbd. is this needed?
} );
...

useFirestoreEmulator(firestore, 'localhost',FIRESTORE_PORT);
useFunctionsEmulator(fns, 'localhost',FUNCTIONS_PORT);
useAuthEmulator(auth, AUTH_URL);

...
await signInWithCustomToken( auth, JSON.stringify({ uid: "dev" }) );   // GETS STUCK!!
```

The promise never resolves. No console output, no error.

firebase-tools 9.5.0
firebase 0.900.15

**Comment by yuchenshi** [comment, marked as answer]
Weird, I just tried this and it signed me in correctly. It's also weird that you're not seeing any console output. Would you mind looking into the Network tab in your DevTools and see what request it is sending? Pay special attention to the signin request and see which host and port it is sending the request to. Also, which browser are you using and what's the format for `AUTH_URL`?

**Reply by akauppi** [reply]
Here are the fetches in the networks tab:

![image](https://user-images.githubusercontent.com/1297734/109365530-1191da00-789a-11eb-9712-a9aca9a11be0.png)

One `OPTIONS` to `accounts:signInWithCustomToken?...`, followed by a `POST` that has a good JWT response.

One `OPTIONS` to `accounts:lookup?key=none`, followed by a `POST` to same address. Great payload.

I've configured the auth port to 9100 - but that's not the cause (tried with default 9099, too).

Browser: Chrome 88 (Safari behaved the same)
macOS 11.2

`AUTH_URL` is "http://localhost:9100"

---

If you wish to try to reproduce:

```
$ git clone git@github.com:akauppi/GroundLevel-es-firebase.git
$ cd GroundLevel*
$ git checkout recamp-firebase-9x
$ npm install
$ cd packages/app
$ npm run dev
```

Open `http://localhost:3000?user=dev`

**Reply by yuchenshi** [reply]
Thanks! We're aware of an issue that may cause the new SDK to get stuck on initialization and we're looking into fixes. Thanks again for bringing this to our attention and please stay tuned -- a quick fix may be coming.

**Reply by sam-gc** [reply]
This issue should be fixed in the next release (https://github.com/firebase/firebase-js-sdk/pull/4550)

Edit: although this issue will still crop up in Safari, as Safari initializes the iframe early; we also need to update the emulator code, which is in the works

**Reply by akauppi** [reply]
I confirm that `signInWithCustomToken` works in 0.900.16 (Chrome). It also seems to work for me in Safari (14.0.3)
